File: backend/src/models/pipeline.py
import uuid
from datetime import datetime
from enum import Enum
from typing import Optional

from pydantic import BaseModel, Field


class AgentType(str, Enum):
    BUILD = "build"
    TEST = "test"
    SECURITY = "security"
    DEPLOY = "deploy"
    VERIFY = "verify"


class Stage(BaseModel):
    id: str
    agent: AgentType
    command: str
    depends_on: list[str] = []
    timeout_seconds: int = 300
    retry_count: int = 0
    critical: bool = True
    env_vars: dict[str, str] = {}


class RepoAnalysis(BaseModel):
    language: str
    framework: Optional[str] = None
    package_manager: str
    has_dockerfile: bool = False
    has_tests: bool = False
    test_runner: Optional[str] = None
    is_monorepo: bool = False


class PipelineSpec(BaseModel):
    pipeline_id: str = Field(default_factory=lambda: str(uuid.uuid4()))
    repo_url: str = ""
    goal: str = ""
    created_at: datetime = Field(default_factory=datetime.utcnow)
    analysis: RepoAnalysis
    stages: list[Stage]

File: frontend/src/types/pipeline.ts
export type AgentType = 'build' | 'test' | 'security' | 'deploy' | 'verify';

export type StageStatus = 'pending' | 'running' | 'success' | 'failed' | 'skipped';

export interface Stage {
  id: string;
  agent: AgentType;
  command: string;
  depends_on: string[];
  timeout_seconds: number;
  retry_count: number;
  critical: boolean;
  env_vars: Record<string, string>;
}

export interface RepoAnalysis {
  language: string;
  framework: string | null;
  package_manager: string;
  has_dockerfile: boolean;
  has_tests: boolean;
  test_runner: string | null;
  is_monorepo: boolean;
}

export interface PipelineSpec {
  pipeline_id: string;
  name: string;
  repo_url: string;
  goal: string;
  created_at: string;
  analysis: RepoAnalysis;
  stages: Stage[];
}

export interface StageResult {
  stage_id: string;
  status: StageStatus;
  exit_code: number;
  stdout: string;
  stderr: string;
  duration_seconds: number;
  artifacts: string[];
  metadata: Record<string, unknown>;
}

export interface StageUpdate {
  stage_id: string;
  status: StageStatus;
  duration_seconds?: number;
  log_tail?: string;
  recovery_strategy?: string;
  recovery_reason?: string;
  modified_command?: string;
  log_type?: LogType;
  log_message?: string;
}

export type LogType =
  | 'stage_start'
  | 'stage_success'
  | 'stage_failed'
  | 'stage_skipped'
  | 'retry'
  | 'recovery_start'
  | 'recovery_plan'
  | 'recovery_success'
  | 'recovery_failed'
  | 'pipeline_start'
  | 'pipeline_done'
  | 'info';

export interface LogEntry {
  timestamp: string;
  stage_id?: string;
  type: LogType;
  message: string;
  details?: string;
}

export interface HistoryEntry {
  pipeline: PipelineSpec;
  results: Record<string, StageResult> | null;
  completedAt: string;
  overallStatus: 'success' | 'failed' | 'partial';
}
